core/shared/src/main/scala/org/http4s/Query.scala

```diff
@@ -171,24 +146,123 @@ final class Query private (value: Either[Vector[KeyValue], String])
 }
 
 object Query {
-  type KeyValue = (String, Option[String])
+  sealed abstract class Component extends Renderable {
+    def separator: Char = '&'
+    def key: String
+    def value: Option[String]
+
+    def normalize: Component
+
+    override def equals(that: Any): Boolean = that match {
+      case that: Component => that.renderString == renderString
+      case _ => false
+    }
+
+    override def hashCode = renderString.##
+  }
+  object Component {
+    sealed abstract class KeyValue extends Component {
+      override def value: Some[String]
+      override def normalize: KeyValue
+    }
+
+    object KeyValue {
+      def apply(key: String, value: String): KeyValue = new KeyValueDecoded(key, value)
+    }
+    sealed abstract class KeyOnly extends Component {
+      override val value: None.type = None
+      override def normalize: KeyOnly
+    }
+
+    object KeyOnly {
+      def apply(name: String): KeyOnly = new KeyOnlyDecoded(name)
+      def unapply(wv: KeyOnly): Some[String] = Some(wv.key)
+    }
```

Comment:
As far as I can see from [RFC 3987](https://www.rfc-editor.org/rfc/rfc3987.html), dividing query components into key and value is just a common use case rather than something built into the query spec - so it isn't really accurate to think of a non-key-value component as having a key but no value. Would it be too much of a departure to change that terminology?